### PR TITLE
Refactoring for multiple layers

### PR DESCRIPTION
**Pull request by subutai**

### CODE REVIEW

**Comment by subutai** [comment]
@lscheinkman Thoughts? This approach (not fully implemented yet) should make it much easier to implement multiple layers and use our SDR logic in many places.

File: htmresearch/frameworks/pytorch/sparse_mnist_net.py [hunk 1]
```diff
@@ -74,80 +72,43 @@ def __init__(self, n=2000,
...
     print("boostStrength is now:", self.boostStrength)
 
 
   def forward(self, x):
-
-    if not self.training:
-      k = min(int(round(self.k * self.kInferenceFactor)), self.n)
-    else:
-      k = self.k
+    self.learningIterations += x.shape[0]
```
Comment (lscheinkman) [inline htmresearch/frameworks/pytorch/sparse_mnist_net.py]: `learningIterations` is not used in `SparseMNISTNet`

File: htmresearch/frameworks/pytorch/linear_sdr.py [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+    if self.weightSparsity < 1.0:
+      self.l1.weight.data[self.zeroWts] = 0.0
+
+
+  def forward(self, x):
+
+    if not self.training:
+      k = min(int(round(self.k * self.kInferenceFactor)), self.n)
+    else:
+      k = self.k
```
Comment (lscheinkman) [inline htmresearch/frameworks/pytorch/linear_sdr.py]: We could call `rezeroWeights`  here and remove the call from the outer models

File: htmresearch/frameworks/pytorch/sparse_mnist_net.py [hunk 3]
```diff
@@ -74,80 +72,43 @@ def __init__(self, n=2000,
...
-      self.rezeroWeights()
+    self.rezeroWeights()
 
 
   def rezeroWeights(self):
-    if self.weightSparsity < 1.0:
-      # print("non zero before:",self.l1.weight.data.nonzero().shape)
-      self.l1.weight.data[self.zeroWts] = 0.0
-      # print("non zero after:",self.l1.weight.data.nonzero().shape)
+    self.l1.rezeroWeights()
```
Comment (lscheinkman) [inline htmresearch/frameworks/pytorch/sparse_mnist_net.py]: Should be moved to `LinearSDR` forward method. See previous comment

**Comment by subutai** [comment]
@lscheinkman Here's a more complete version, takes your comments into account. Doing some testing now, before merging to master.

File: htmresearch/frameworks/pytorch/sparse_mnist_cnn.py [hunk 4]
```diff
@@ -88,8 +51,8 @@ def __init__(self,
 
     :param c1k:
       Number of ON (non-zero) units per iteration in the first convolutional
-      layer C1. The sparsity of this layer will be c1k / self.c1OutputLength.
-      If c1k >= self.c1OutputLength, the layer acts as a traditional
+      layer C1. The sparsity of this layer will be c1k / self.c1.outputLength.
+      If c1k >= self.c1.outputLength, the layer acts as a traditional
```
Comment (lscheinkman) [inline htmresearch/frameworks/pytorch/sparse_mnist_cnn.py]: `self.c1` is now called `self.cnnSdr1`

Comment (subutai) [reply htmresearch/frameworks/pytorch/sparse_mnist_cnn.py]: Fixed

**Pull request merged**